pulp_ansible/app/tasks/signature.py

```diff
@@ -0,0 +1,91 @@
+)
+
+from pulp_ansible.app.models import (
+    AnsibleRepository,
+    CollectionVersion,
+    CollectionVersionSignature,
+)
+
+from pulpcore.plugin.models import Artifact, SigningService
+from pulpcore.plugin.sync import sync_to_async_iterable
+
+
+def sign(repository_href, content_hrefs, signing_service_href):
+    """The signing task."""
+    repository = AnsibleRepository.objects.get(pk=repository_href)
+    if content_hrefs == ["*"]:
+        content = repository.latest_version().content.filter(pulp_type=CollectionVersion.get_pulp_type())
+    else:
+        content = CollectionVersion.objects.filter(pk__in=content_hrefs)
+    signing_service = SigningService.objects.get(pk=signing_service_href)
+    first_stage = CollectionSigningStage(content, signing_service)
+    SigningDeclarativeVersion(first_stage, repository).create()
+
+
+class SigningDeclarativeVersion(DeclarativeVersion):
+    """Custom signature pipeline."""
+
+    def pipeline_stages(self, new_version):
+        """The stages for the signing process."""
+        pipeline = [
+            self.first_stage,  # CollectionSigningStage
+            QueryExistingArtifacts(),
+            ArtifactSaver(),
+            # QueryExistingContents(),
+            ContentSaver(),
+        ]
+        return pipeline
+
+
+class CollectionSigningStage(Stage):
+    """
+    This stage signs the content and creates CollectionVersionSignatures.
+    """
+
+    def __init__(self, content, signing_service):
+        super().__init__()
+        self.content = content
+        self.signing_service = signing_service
+        self.base_dir = os.getcwd()
+
+    async def run(self):
+        """does the magic."""
+        async for collection_version in sync_to_async_iterable(self.content.iterator()):
+            # Can add extra parallelization here with tasks
+            # We use the manifest to create the signature
+            d = tempfile.mkdtemp(dir=self.base_dir)
+            os.chdir(d)  # GPG doesn't take filename into account for signatures, but not sure about other methods
+            with open(f"MANIFEST.json", 'w') as t:
+                json.dump(collection_version.manifest, t)
+
+            # Signatures typically have their date created embedded within the signature which means multiple
+            # signatures on the same content from the same signing-service will produce multiple valid and
+            # *unique* signatures which won't be de-duplicated. Maybe another setting should be added to act
+            # as a policy for these potential "duplicate" signatures.
+            result = await self.signing_service.asign(t.name)
+            # Is it always a guarantee that the result will have a "signature" where the file is?
+            artifact = Artifact.init_and_validate(result["signature"])
+            # Should this relative path be the name of the "{collection-version}-signature"?
+            da = DeclarativeArtifact(artifact, url="NA", relative_path="MANIFEST.json.asc")
```

Comment:
I mentioned on the call, but wanted to leave it here that I think we should be storing this signature entirely in memory and there would be no `Artifact` or `DeclarativeArtifact` here. Especially given that we don't want the API to have to read the Artifact data later from S3 or Azure object store backends.

Reply:
Sure, I'll change the model to have the signature be apart of it.